File: internal/domain/catalog.go
package domain

import (
	"fmt"
	"regexp"
	"sort"
	"strings"
)

// MetricSpec pins everything about one CloudWatch metric in a single place:
// which namespace and statistic it is read with, what unit it carries, and how
// it is coloured.
//
// The statistic in particular belongs here rather than at each call site. Read
// the same counter as Sum on one screen and Average on another and the two
// screens disagree forever, with nothing in either one hinting at why.
type MetricSpec struct {
	Key        string
	Label      string
	Namespace  string
	MetricName string
	Stat       string
	Unit       Unit
	Color      string
	Intent     Intent

	// Dimensions are the dimension names the SEARCH expression groups by. The
	// last one is what distinguishes the individual series.
	Dimensions []string
}

// CloudWatch namespaces.
const (
	NSApplicationELB  = "AWS/ApplicationELB"
	NSContainer       = "ContainerInsights"
	NSRDS             = "AWS/RDS"
	NSWAFV2           = "AWS/WAFV2"
	NSContainerPodSvc = "ContainerInsights"
)

// Statistics. Counters are summed, utilisations are averaged, latencies are
// read at percentiles.
const (
	StatSum = "Sum"
	StatAvg = "Average"
	StatMax = "Maximum"
	StatMin = "Minimum"
	StatP50 = "p50"
	StatP90 = "p90"
	StatP99 = "p99"
)

// Apple's system palette, referenced by name. The CSS custom properties that
// carry the actual oklch values live in the frontend; the backend only decides
// which semantic colour a series should take, so a metric's meaning and its
// colour cannot drift apart between panels.
const (
	ColorBlue   = "systemBlue"
	ColorGreen  = "systemGreen"
	ColorIndigo = "systemIndigo"
	ColorOrange = "systemOrange"
	ColorPink   = "systemPink"
	ColorPurple = "systemPurple"
	ColorRed    = "systemRed"
	ColorTeal   = "systemTeal"
	ColorYellow = "systemYellow"
	ColorMint   = "systemMint"
	ColorGray   = "systemGray"
)

// SubjectPalette colours one subject — a pod, a node — on a panel where a
// single spec fans out into many series.
//
// A spec's own colour says what the line measures, which is the right answer
// when a chart holds one line per metric. It is the wrong answer when the
// SEARCH fans out: every pod on the CPU panel took ColorIndigo, so twenty pods
// drew twenty identical lines and the legend text was the only way to tell them
// apart. Colour has to carry the subject there, and the metric moves to the
// line's dash pattern — see VariantDash.
//
// The order is not the constant block's. Neighbouring entries are kept apart
// under red-green and blue-yellow colour vision deficiency, so a chart with
// three pods on it does not hand out three colours that some readers see as
// one. ColorGray is excluded: sumSeries already spends it on totals.
var SubjectPalette = []string{
	ColorBlue, ColorOrange, ColorGreen, ColorPink, ColorPurple,
	ColorYellow, ColorTeal, ColorRed, ColorIndigo, ColorMint,
}

// SubjectColor picks the palette entry for the i-th subject, cycling when a
// panel holds more subjects than the palette has colours. Past that point the
// dash pattern and the legend label are what separate two lines; the caller is
// expected to order subjects deterministically so a colour does not move
// between refreshes.
func SubjectColor(i int) string {
	if i < 0 {
		i = 0
	}
	return SubjectPalette[i%len(SubjectPalette)]
}

// TargetGroupMetrics covers requirement 3: target response time, 5xx and 4xx.
func TargetGroupMetrics() []MetricSpec {
	dims := []string{"LoadBalancer", "TargetGroup"}
	return []MetricSpec{
		{Key: "tg.p50", Label: "응답 시간 p50", Namespace: NSApplicationELB, MetricName: "TargetResponseTime", Stat: StatP50, Unit: UnitSeconds, Color: ColorTeal, Dimensions: dims},
		{Key: "tg.p90", Label: "응답 시간 p90", Namespace: NSApplicationELB, MetricName: "TargetResponseTime", Stat: StatP90, Unit: UnitSeconds, Color: ColorBlue, Dimensions: dims},
		{Key: "tg.p99", Label: "응답 시간 p99", Namespace: NSApplicationELB, MetricName: "TargetResponseTime", Stat: StatP99, Unit: UnitSeconds, Color: ColorIndigo, Dimensions: dims},
		{Key: "tg.4xx", Label: "대상 4xx", Namespace: NSApplicationELB, MetricName: "HTTPCode_Target_4XX_Count", Stat: StatSum, Unit: UnitCount, Color: ColorOrange, Intent: IntentWarn, Dimensions: dims},
		{Key: "tg.5xx", Label: "대상 5xx", Namespace: NSApplicationELB, MetricName: "HTTPCode_Target_5XX_Count", Stat: StatSum, Unit: UnitCount, Color: ColorRed, Intent: IntentBad, Dimensions: dims},
		{Key: "tg.requests", Label: "요청 수", Namespace: NSApplicationELB, MetricName: "RequestCount", Stat: StatSum, Unit: UnitCount, Color: ColorGray, Dimensions: dims},
		{Key: "tg.healthy", Label: "정상 대상", Namespace: NSApplicationELB, MetricName: "HealthyHostCount", Stat: StatAvg, Unit: UnitCount, Color: ColorGreen, Intent: IntentGood, Dimensions: dims},
	}
}

// PodResourceMetrics covers requirement 4.
func PodResourceMetrics() []MetricSpec {
	dims := []string{"ClusterName", "Namespace", "PodName"}
	return []MetricSpec{
		{Key: "pod.cpu", Label: "CPU 사용률", Namespace: NSContainer, MetricName: "pod_cpu_utilization", Stat: StatAvg, Unit: UnitPercent, Color: ColorIndigo, Dimensions: dims},
		{Key: "pod.mem", Label: "메모리 사용률", Namespace: NSContainer, MetricName: "pod_memory_utilization", Stat: StatAvg, Unit: UnitPercent, Color: ColorTeal, Dimensions: dims},
		{Key: "pod.cpu.limit", Label: "CPU 사용률 (limit 대비)", Namespace: NSContainer, MetricName: "pod_cpu_utilization_over_pod_limit", Stat: StatAvg, Unit: UnitPercent, Color: ColorPurple, Dimensions: dims},
		{Key: "pod.mem.limit", Label: "메모리 사용률 (limit 대비)", Namespace: NSContainer, MetricName: "pod_memory_utilization_over_pod_limit", Stat: StatAvg, Unit: UnitPercent, Color: ColorPink, Intent: IntentWarn, Dimensions: dims},
	}
}

// NodeResourceMetrics covers requirement 5.
//
// InstanceId belongs in the schema even though nothing pins it: Container
// Insights publishes the node metrics under {NodeName, ClusterName, InstanceId}
// and {ClusterName}, and nothing else. A SEARCH schema is matched as a set, so
// {ClusterName,NodeName} matched no metric at all and the panel rendered empty
// — indistinguishable from a cluster with Container Insights switched off.
func NodeResourceMetrics() []MetricSpec {
	dims := []string{"ClusterName", "InstanceId", "NodeName"}
	return []MetricSpec{
		{Key: "node.cpu", Label: "CPU 사용률", Namespace: NSContainer, MetricName: "node_cpu_utilization", Stat: StatAvg, Unit: UnitPercent, Color: ColorIndigo, Dimensions: dims},
		{Key: "node.mem", Label: "메모리 사용률", Namespace: NSContainer, MetricName: "node_memory_utilization", Stat: StatAvg, Unit: UnitPercent, Color: ColorTeal, Dimensions: dims},
		{Key: "node.fs", Label: "디스크 사용률", Namespace: NSContainer, MetricName: "node_filesystem_utilization", Stat: StatAvg, Unit: UnitPercent, Color: ColorYellow, Dimensions: dims},
	}
}

// CountMetrics covers requirement 6: how many pods and nodes there are.
func CountMetrics() []MetricSpec {
	return []MetricSpec{
		{Key: "count.pods", Label: "실행 중 팟", Namespace: NSContainer, MetricName: "service_number_of_running_pods", Stat: StatAvg, Unit: UnitCount, Color: ColorBlue, Dimensions: []string{"ClusterName", "Namespace", "Service"}},
		{Key: "count.nodes", Label: "노드", Namespace: NSContainer, MetricName: "cluster_node_count", Stat: StatAvg, Unit: UnitCount, Color: ColorGreen, Dimensions: []string{"ClusterName"}},
		{Key: "count.nodes.failed", Label: "실패 노드", Namespace: NSContainer, MetricName: "cluster_failed_node_count", Stat: StatMax, Unit: UnitCount, Color: ColorRed, Intent: IntentBad, Dimensions: []string{"ClusterName"}},
	}
}

// PodStatusMetrics covers requirement 7.
//
// PodName is in the schema because none of these metrics is published at
// {Namespace, ClusterName}: the pod-level set is {PodName, Namespace,
// ClusterName}. The panel sums the fan-out back down, so the numbers mean the
// same thing they were meant to; without PodName they meant nothing, because
// the SEARCH matched no series.
//
// pod_status_* is published only by Container Insights with enhanced
// observability. pod_number_of_container_restarts is published by both, which
// is why the panel can still say something on a cluster running the older
// agent.
func PodStatusMetrics() []MetricSpec {
	dims := []string{"ClusterName", "Namespace", "PodName"}
	return []MetricSpec{
		{Key: "pod.running", Label: "Running", Namespace: NSContainer, MetricName: "pod_status_running", Stat: StatAvg, Unit: UnitCount, Color: ColorGreen, Intent: IntentGood, Dimensions: dims},
		{Key: "pod.pending", Label: "Pending", Namespace: NSContainer, MetricName: "pod_status_pending", Stat: StatAvg, Unit: UnitCount, Color: ColorOrange, Intent: IntentWarn, Dimensions: dims},
		{Key: "pod.failed", Label: "Failed", Namespace: NSContainer, MetricName: "pod_status_failed", Stat: StatAvg, Unit: UnitCount, Color: ColorRed, Intent: IntentBad, Dimensions: dims},
		{Key: "pod.restarts", Label: "컨테이너 재시작", Namespace: NSContainer, MetricName: "pod_number_of_container_restarts", Stat: StatSum, Unit: UnitCount, Color: ColorPink, Intent: IntentBad, Dimensions: dims},
	}
}

// RDSProxyMetrics covers requirement 8.
func RDSProxyMetrics() []MetricSpec {
	dims := []string{"ProxyName"}
	return []MetricSpec{
		{Key: "proxy.db", Label: "DB 커넥션", Namespace: NSRDS, MetricName: "DatabaseConnections", Stat: StatAvg, Unit: UnitConns, Color: ColorPurple, Dimensions: dims},
		{Key: "proxy.client", Label: "클라이언트 커넥션", Namespace: NSRDS, MetricName: "ClientConnections", Stat: StatAvg, Unit: UnitConns, Color: ColorBlue, Dimensions: dims},
		{Key: "proxy.pinned", Label: "세션 고정 커넥션", Namespace: NSRDS, MetricName: "DatabaseConnectionsCurrentlySessionPinned", Stat: StatAvg, Unit: UnitConns, Color: ColorOrange, Intent: IntentWarn, Dimensions: dims},
		{Key: "proxy.borrow", Label: "커넥션 대기", Namespace: NSRDS, MetricName: "DatabaseConnectionsBorrowLatency", Stat: StatAvg, Unit: UnitMillis, Color: ColorYellow, Dimensions: dims},
		{Key: "proxy.max", Label: "최대 허용 커넥션", Namespace: NSRDS, MetricName: "MaxDatabaseConnectionsAllowed", Stat: StatMax, Unit: UnitConns, Color: ColorGray, Dimensions: dims},
	}
}

// WAFMetrics is the metric-side view of the WAF, complementing the log-derived
// statistics on the same page.
func WAFMetrics() []MetricSpec {
	dims := []string{"WebACL", "Rule", "Region"}
	return []MetricSpec{
		// No intent on either: a WAF blocking requests is the normal operating
		// state, so flagging it permanently lit the alarm and buried the cases
		// that actually deviate. Colour still separates the two.
		{Key: "waf.allowed", Label: "허용", Namespace: NSWAFV2, MetricName: "AllowedRequests", Stat: StatSum, Unit: UnitCount, Color: ColorGreen, Dimensions: dims},
		{Key: "waf.blocked", Label: "차단", Namespace: NSWAFV2, MetricName: "BlockedRequests", Stat: StatSum, Unit: UnitCount, Color: ColorPink, Dimensions: dims},
		{Key: "waf.counted", Label: "카운트", Namespace: NSWAFV2, MetricName: "CountedRequests", Stat: StatSum, Unit: UnitCount, Color: ColorGray, Dimensions: dims},
	}
}

// SpecsWithPrefix picks the specs whose key begins with prefix, so a panel can
// plot CPU without memory sharing its chart. The keys already nest — pod.cpu
// carries pod.cpu.limit with it — so one prefix names a whole resource.
func SpecsWithPrefix(specs []MetricSpec, prefix string) []MetricSpec {
	out := make([]MetricSpec, 0, len(specs))
	for _, s := range specs {
		if strings.HasPrefix(s.Key, prefix) {
			out = append(out, s)
		}
	}
	return out
}

// AllMetrics is every spec the dashboard knows about, used to assert that keys
// are unique and that no metric is read two different ways.
func AllMetrics() []MetricSpec {
	var out []MetricSpec
	for _, group := range [][]MetricSpec{
		TargetGroupMetrics(), PodResourceMetrics(), NodeResourceMetrics(),
		CountMetrics(), PodStatusMetrics(), RDSProxyMetrics(), WAFMetrics(),
	} {
		out = append(out, group...)
	}
	return out
}

// searchValueRe keeps a dimension value to characters that are safe inside the
// double-quoted term of a SEARCH expression.
var searchValueRe = regexp.MustCompile(`^[A-Za-z0-9 ._:/+=@-]+$`)

// SearchExpression renders a CloudWatch metric-math SEARCH for spec, filtered
// by the given dimension values.
//
// Using SEARCH removes the ListMetrics call entirely, and with it the failure
// that eventually took the reference implementation's metric panels offline:
// ListMetrics keeps returning dimensions for pods that no longer exist, so the
// generated query list grew monotonically until it crossed GetMetricData's
// 500-query ceiling and every subsequent call failed validation. A SEARCH is
// one query no matter how many series it matches.
func SearchExpression(spec MetricSpec, filters map[string]string, p Period) (string, error) {
	if spec.Namespace == "" || spec.MetricName == "" {
		return "", fmt.Errorf("metric %q is missing a namespace or name", spec.Key)
	}
	schema := append([]string{spec.Namespace}, spec.Dimensions...)

	terms := []string{fmt.Sprintf("MetricName=%q", spec.MetricName)}
	keys := make([]string, 0, len(filters))
	for k := range filters {
		keys = append(keys, k)
	}
	sort.Strings(keys) // deterministic output, so queries cache and tests compare
	for _, k := range keys {
		v := filters[k]
		if v == "" {
			continue
		}
		if !searchValueRe.MatchString(k) || !searchValueRe.MatchString(v) {
			return "", fmt.Errorf("dimension %s=%q contains characters that cannot be embedded in a SEARCH expression", k, v)
		}
		terms = append(terms, fmt.Sprintf("%s=%q", k, v))
	}

	return fmt.Sprintf("SEARCH('{%s} %s', '%s', %d)",
		strings.Join(schema, ","),
		strings.Join(terms, " "),
		spec.Stat,
		p.Seconds(),
	), nil
}

// queryIDRe matches what GetMetricData accepts as a query identifier.
var queryIDRe = regexp.MustCompile(`^[a-z][a-zA-Z0-9_]*$`)

// QueryID turns a metric key into a valid GetMetricData id.
//
// Results are matched back to specs through this identifier rather than by
// position. The reference implementation recovered an array index by scanning
// the id with Sscanf, discarded the error, and then indexed a slice with the
// result — so an unexpected id wrote data into the wrong series, and an
// out-of-range one panicked the handler.
func QueryID(key string) string {
	var b strings.Builder
	b.WriteByte('q')
	for _, r := range key {
		switch {
		case r >= 'a' && r <= 'z', r >= 'A' && r <= 'Z', r >= '0' && r <= '9':
			b.WriteRune(r)
		default:
			b.WriteByte('_')
		}
	}
	id := b.String()
	if !queryIDRe.MatchString(id) {
		// Unreachable for any key made of printable characters, but a
		// malformed id would be rejected by CloudWatch rather than silently
		// mismatched, so fall back to something valid.
		return "q"
	}
	return id
}

File: server/src/domain/catalog.ts
// 이 대시보드가 아는 CloudWatch 메트릭 전부. internal/domain/catalog.go 의
// 이식이다.

import type { Intent, Unit } from '../contract.ts';

/**
 * MetricSpec 은 메트릭 하나에 대한 모든 것을 한자리에 못박는다. 어느 네임스페이스와
 * 통계로 읽는지, 어떤 단위를 지니는지, 어떤 색인지.
 *
 * 특히 통계가 호출 지점이 아니라 여기 있어야 한다. 같은 카운터를 한 화면에서는
 * Sum 으로, 다른 화면에서는 Average 로 읽으면 두 화면은 영원히 어긋나고, 어느
 * 쪽에도 그 이유를 짐작할 단서가 없다.
 */
export interface MetricSpec {
	key: string;
	label: string;
	namespace: string;
	metricName: string;
	stat: string;
	unit: Unit;
	color: string;
	intent?: Intent;
	/** SEARCH 식이 묶는 차원 이름. 마지막 것이 개별 시리즈를 가른다. */
	dimensions: string[];
}

// CloudWatch 네임스페이스.
export const nsApplicationELB = 'AWS/ApplicationELB';
export const nsContainer = 'ContainerInsights';
export const nsRDS = 'AWS/RDS';
export const nsWAFV2 = 'AWS/WAFV2';

// 통계. 카운터는 더하고, 사용률은 평균하고, 지연은 백분위로 읽는다.
export const statSum = 'Sum';
export const statAvg = 'Average';
export const statMax = 'Maximum';
export const statMin = 'Minimum';
export const statP50 = 'p50';
export const statP90 = 'p90';
export const statP99 = 'p99';

// 애플 시스템 팔레트를 이름으로 참조한다. 실제 oklch 값을 담은 CSS 변수는
// 프론트에 있다. 백엔드는 어떤 의미의 색을 쓸지만 정하므로, 메트릭의 뜻과 색이
// 패널 사이에서 어긋날 수 없다.
export const colorBlue = 'systemBlue';
export const colorGreen = 'systemGreen';
export const colorIndigo = 'systemIndigo';
export const colorOrange = 'systemOrange';
export const colorPink = 'systemPink';
export const colorPurple = 'systemPurple';
export const colorRed = 'systemRed';
export const colorTeal = 'systemTeal';
export const colorYellow = 'systemYellow';
export const colorMint = 'systemMint';
export const colorGray = 'systemGray';

/**
 * subjectPalette 는 스펙 하나가 여러 시리즈로 퍼지는 패널에서 주체 하나 — 팟,
 * 노드 — 에 색을 준다.
 *
 * 스펙 자신의 색은 그 선이 무엇을 재는지를 말하고, 차트가 메트릭당 선 하나를
 * 들 때는 그게 옳은 답이다. SEARCH 가 퍼질 때는 틀린 답이 된다 — CPU 패널의 모든
 * 팟이 colorIndigo 를 가져가서 팟 스무 개가 똑같은 선 스무 개를 그렸고, 범례
 * 글자만이 그것을 가르는 유일한 수단이었다. 거기서는 색이 주체를 실어야 하고,
 * 메트릭은 선 무늬로 옮겨 간다 — variantDash 참고.
 *
 * 순서는 상수 블록의 순서가 아니다. 이웃한 항목이 적록·청황 색각 이상에서도
 * 갈라지도록 배치했다. 팟 셋이 올라간 차트가 어떤 독자에게는 하나로 보이는 색
 * 셋을 내주지 않게 하려는 것이다. colorGray 는 뺐다 — sumSeries 가 이미 합계에
 * 그 색을 쓴다.
 */
export const subjectPalette = [
	colorBlue,
	colorOrange,
	colorGreen,
	colorPink,
	colorPurple,
	colorYellow,
	colorTeal,
	colorRed,
	colorIndigo,
	colorMint
];

/**
 * subjectColor 는 i 번째 주체의 팔레트 항목을 고르고, 팔레트보다 주체가 많으면
 * 돌려 쓴다. 그 뒤로는 선 무늬와 범례가 두 선을 가른다. 호출자는 주체를 결정적인
 * 순서로 늘어놓아 새로고침 때마다 색이 옮겨 다니지 않게 해야 한다.
 */
export function subjectColor(i: number): string {
	const n = i < 0 ? 0 : i;
	return subjectPalette[n % subjectPalette.length] as string;
}

/** 요구사항 3: 대상 응답 시간, 5xx, 4xx. */
export function targetGroupMetrics(): MetricSpec[] {
	const dimensions = ['LoadBalancer', 'TargetGroup'];
	return [
		{ key: 'tg.p50', label: '응답 시간 p50', namespace: nsApplicationELB, metricName: 'TargetResponseTime', stat: statP50, unit: 's', color: colorTeal, dimensions },
		{ key: 'tg.p90', label: '응답 시간 p90', namespace: nsApplicationELB, metricName: 'TargetResponseTime', stat: statP90, unit: 's', color: colorBlue, dimensions },
		{ key: 'tg.p99', label: '응답 시간 p99', namespace: nsApplicationELB, metricName: 'TargetResponseTime', stat: statP99, unit: 's', color: colorIndigo, dimensions },
		{ key: 'tg.4xx', label: '대상 4xx', namespace: nsApplicationELB, metricName: 'HTTPCode_Target_4XX_Count', stat: statSum, unit: 'count', color: colorOrange, intent: 'warn', dimensions },
		{ key: 'tg.5xx', label: '대상 5xx', namespace: nsApplicationELB, metricName: 'HTTPCode_Target_5XX_Count', stat: statSum, unit: 'count', color: colorRed, intent: 'bad', dimensions },
		{ key: 'tg.requests', label: '요청 수', namespace: nsApplicationELB, metricName: 'RequestCount', stat: statSum, unit: 'count', color: colorGray, dimensions },
		{ key: 'tg.healthy', label: '정상 대상', namespace: nsApplicationELB, metricName: 'HealthyHostCount', stat: statAvg, unit: 'count', color: colorGreen, intent: 'good', dimensions }
	];
}

/** 요구사항 4. */
export function podResourceMetrics(): MetricSpec[] {
	const dimensions = ['ClusterName', 'Namespace', 'PodName'];
	return [
		{ key: 'pod.cpu', label: 'CPU 사용률', namespace: nsContainer, metricName: 'pod_cpu_utilization', stat: statAvg, unit: '%', color: colorIndigo, dimensions },
		{ key: 'pod.mem', label: '메모리 사용률', namespace: nsContainer, metricName: 'pod_memory_utilization', stat: statAvg, unit: '%', color: colorTeal, dimensions },
		{ key: 'pod.cpu.limit', label: 'CPU 사용률 (limit 대비)', namespace: nsContainer, metricName: 'pod_cpu_utilization_over_pod_limit', stat: statAvg, unit: '%', color: colorPurple, dimensions },
		{ key: 'pod.mem.limit', label: '메모리 사용률 (limit 대비)', namespace: nsContainer, metricName: 'pod_memory_utilization_over_pod_limit', stat: statAvg, unit: '%', color: colorPink, intent: 'warn', dimensions }
	];
}

/**
 * 요구사항 5.
 *
 * 아무것도 고정하지 않는데도 InstanceId 가 스키마에 있어야 한다. Container
 * Insights 는 노드 메트릭을 {NodeName, ClusterName, InstanceId} 와 {ClusterName}
 * 아래에만 publish 한다. SEARCH 스키마는 집합으로 맞춰지므로 {ClusterName,NodeName}
 * 은 어떤 메트릭에도 맞지 않았고 패널은 비어서 떴다 — Container Insights 가 꺼진
 * 클러스터와 구분되지 않는 모습으로.
 */
export function nodeResourceMetrics(): MetricSpec[] {
	const dimensions = ['ClusterName', 'InstanceId', 'NodeName'];
	return [
		{ key: 'node.cpu', label: 'CPU 사용률', namespace: nsContainer, metricName: 'node_cpu_utilization', stat: statAvg, unit: '%', color: colorIndigo, dimensions },
		{ key: 'node.mem', label: '메모리 사용률', namespace: nsContainer, metricName: 'node_memory_utilization', stat: statAvg, unit: '%', color: colorTeal, dimensions },
		{ key: 'node.fs', label: '디스크 사용률', namespace: nsContainer, metricName: 'node_filesystem_utilization', stat: statAvg, unit: '%', color: colorYellow, dimensions }
	];
}

/** 요구사항 6: 팟과 노드가 몇 개인가. */
export function countMetrics(): MetricSpec[] {
	return [
		{ key: 'count.pods', label: '실행 중 팟', namespace: nsContainer, metricName: 'service_number_of_running_pods', stat: statAvg, unit: 'count', color: colorBlue, dimensions: ['ClusterName', 'Namespace', 'Service'] },
		{ key: 'count.nodes', label: '노드', namespace: nsContainer, metricName: 'cluster_node_count', stat: statAvg, unit: 'count', color: colorGreen, dimensions: ['ClusterName'] },
		{ key: 'count.nodes.failed', label: '실패 노드', namespace: nsContainer, metricName: 'cluster_failed_node_count', stat: statMax, unit: 'count', color: colorRed, intent: 'bad', dimensions: ['ClusterName'] }
	];
}

/**
 * 요구사항 7.
 *
 * PodName 이 스키마에 있는 이유는 이 메트릭 중 어느 것도 {Namespace, ClusterName}
 * 에 publish 되지 않기 때문이다 — 팟 단위 집합은 {PodName, Namespace, ClusterName}
 * 이다. 패널이 퍼진 것을 다시 더해 내리므로 숫자의 뜻은 원래 의도한 그대로다.
 * PodName 이 없으면 아무 뜻도 없었다. SEARCH 가 어떤 시리즈에도 맞지 않았으니까.
 *
 * pod_status_* 는 향상된 관측이 켜진 Container Insights 만 publish 한다.
 * pod_number_of_container_restarts 는 양쪽 다 내므로, 옛 에이전트를 도는
 * 클러스터에서도 이 패널이 무언가는 말할 수 있다.
 */
export function podStatusMetrics(): MetricSpec[] {
	const dimensions = ['ClusterName', 'Namespace', 'PodName'];
	return [
		{ key: 'pod.running', label: 'Running', namespace: nsContainer, metricName: 'pod_status_running', stat: statAvg, unit: 'count', color: colorGreen, intent: 'good', dimensions },
		{ key: 'pod.pending', label: 'Pending', namespace: nsContainer, metricName: 'pod_status_pending', stat: statAvg, unit: 'count', color: colorOrange, intent: 'warn', dimensions },
		{ key: 'pod.failed', label: 'Failed', namespace: nsContainer, metricName: 'pod_status_failed', stat: statAvg, unit: 'count', color: colorRed, intent: 'bad', dimensions },
		{ key: 'pod.restarts', label: '컨테이너 재시작', namespace: nsContainer, metricName: 'pod_number_of_container_restarts', stat: statSum, unit: 'count', color: colorPink, intent: 'bad', dimensions }
	];
}

/** 요구사항 8. */
export function rdsProxyMetrics(): MetricSpec[] {
	const dimensions = ['ProxyName'];
	return [
		{ key: 'proxy.db', label: 'DB 커넥션', namespace: nsRDS, metricName: 'DatabaseConnections', stat: statAvg, unit: 'conn', color: colorPurple, dimensions },
		{ key: 'proxy.client', label: '클라이언트 커넥션', namespace: nsRDS, metricName: 'ClientConnections', stat: statAvg, unit: 'conn', color: colorBlue, dimensions },
		{ key: 'proxy.pinned', label: '세션 고정 커넥션', namespace: nsRDS, metricName: 'DatabaseConnectionsCurrentlySessionPinned', stat: statAvg, unit: 'conn', color: colorOrange, intent: 'warn', dimensions },
		{ key: 'proxy.borrow', label: '커넥션 대기', namespace: nsRDS, metricName: 'DatabaseConnectionsBorrowLatency', stat: statAvg, unit: 'ms', color: colorYellow, dimensions },
		{ key: 'proxy.max', label: '최대 허용 커넥션', namespace: nsRDS, metricName: 'MaxDatabaseConnectionsAllowed', stat: statMax, unit: 'conn', color: colorGray, dimensions }
	];
}

/** 같은 페이지의 로그 기반 통계를 보완하는, WAF 의 메트릭 쪽 시각. */
export function wafMetrics(): MetricSpec[] {
	const dimensions = ['WebACL', 'Rule', 'Region'];
	return [
		// 둘 다 intent 가 없다. 요청을 차단하는 WAF 는 정상 운영 상태이므로,
		// 그것에 표시를 달면 경보가 영구히 켜져 있고 실제로 벗어난 경우가
		// 묻힌다. 색은 여전히 둘을 가른다.
		{ key: 'waf.allowed', label: '허용', namespace: nsWAFV2, metricName: 'AllowedRequests', stat: statSum, unit: 'count', color: colorGreen, dimensions },
		{ key: 'waf.blocked', label: '차단', namespace: nsWAFV2, metricName: 'BlockedRequests', stat: statSum, unit: 'count', color: colorPink, dimensions },
		{ key: 'waf.counted', label: '카운트', namespace: nsWAFV2, metricName: 'CountedRequests', stat: statSum, unit: 'count', color: colorGray, dimensions }
	];
}

/**
 * specsWithPrefix 는 키가 prefix 로 시작하는 스펙을 고른다. CPU 패널이 메모리와
 * 차트를 나눠 쓰지 않게 하려는 것이다. 키가 이미 중첩돼 있으므로 — pod.cpu 가
 * pod.cpu.limit 을 데려간다 — 접두사 하나가 리소스 하나를 가리킨다.
 */
export function specsWithPrefix(specs: MetricSpec[], prefix: string): MetricSpec[] {
	return specs.filter((s) => s.key.startsWith(prefix));
}

/** 대시보드가 아는 모든 스펙. 키가 유일한지 확인할 때 쓴다. */
export function allMetrics(): MetricSpec[] {
	return [
		...targetGroupMetrics(),
		...podResourceMetrics(),
		...nodeResourceMetrics(),
		...countMetrics(),
		...podStatusMetrics(),
		...rdsProxyMetrics(),
		...wafMetrics()
	];
}

/**
 * searchValueRe 는 차원 값을 SEARCH 식의 큰따옴표 안에 넣어도 안전한 문자로
 * 묶는다.
 */
const searchValueRe = /^[A-Za-z0-9 ._:/+=@-]+$/;

/**
 * searchExpression 은 스펙에 대한 CloudWatch 메트릭 수식 SEARCH 를, 주어진 차원
 * 값으로 걸러서 만든다.
 *
 * SEARCH 를 쓰면 ListMetrics 호출이 통째로 사라지고, 그와 함께 이전 구현의 메트릭
 * 패널을 끝내 죽인 실패도 사라진다. ListMetrics 는 이미 없어진 팟의 차원까지 계속
 * 돌려주므로 생성되는 쿼리 목록이 단조 증가했고, GetMetricData 의 500쿼리 천장을
 * 넘은 순간부터 이후 모든 호출이 검증에서 실패했다. SEARCH 는 몇 개의 시리즈에
 * 맞든 쿼리 하나다.
 */
export function searchExpression(
	spec: MetricSpec,
	filters: Record<string, string>,
	periodSecs: number
): string {
	if (spec.namespace === '' || spec.metricName === '') {
		throw new Error(`metric "${spec.key}" is missing a namespace or name`);
	}
	const schema = [spec.namespace, ...spec.dimensions];

	const terms = [`MetricName=${JSON.stringify(spec.metricName)}`];
	// 키를 정렬한다. 출력이 결정적이어야 쿼리가 캐시되고 테스트가 비교된다.
	for (const k of Object.keys(filters).sort()) {
		const v = filters[k];
		if (v === undefined || v === '') continue;
		if (!searchValueRe.test(k) || !searchValueRe.test(v)) {
			throw new Error(
				`dimension ${k}="${v}" contains characters that cannot be embedded in a SEARCH expression`
			);
		}
		terms.push(`${k}=${JSON.stringify(v)}`);
	}

	return `SEARCH('{${schema.join(',')}} ${terms.join(' ')}', '${spec.stat}', ${periodSecs})`;
}

/** queryIDRe 는 GetMetricData 가 쿼리 식별자로 받아 주는 모양이다. */
const queryIDRe = /^[a-z][a-zA-Z0-9_]*$/;

/**
 * queryID 는 메트릭 키를 쓸 수 있는 GetMetricData id 로 바꾼다.
 *
 * 결과를 스펙에 되맞출 때 위치가 아니라 이 식별자를 쓴다. 이전 구현은 id 를
 * Sscanf 로 훑어 배열 인덱스를 복원하고 오류를 버린 뒤 그 값으로 슬라이스를
 * 인덱싱했다 — 그래서 예상 못 한 id 는 엉뚱한 시리즈에 데이터를 썼고, 범위를
 * 벗어난 id 는 핸들러를 패닉시켰다.
 */
export function queryID(key: string): string {
	let id = 'q';
	for (const ch of key) {
		id += /[A-Za-z0-9]/.test(ch) ? ch : '_';
	}
	// 인쇄 가능한 문자로 된 키에서는 닿지 않는 자리다. 다만 잘못된 id 는
	// 조용히 어긋나는 대신 CloudWatch 에서 거절당하므로, 유효한 것으로 물러난다.
	return queryIDRe.test(id) ? id : 'q';
}
